# Momentum Optimization

Momentum is an optimization technique used with Gradient Descent.

The main problem with normal Gradient Descent is that the updates can
sometimes be:

- slow
- noisy
- oscillatory
- inefficient in certain directions

Momentum solves this by remembering part of the previous update.

Instead of only considering the current gradient, Momentum uses:

    Current Gradient
          +
    Previous Update

This gives the optimizer a kind of "velocity".

## Intuition Behind Momentum

Imagine a ball rolling down a hill.

Without momentum:

    The ball looks only at the current slope.

With momentum:

    The ball remembers its previous movement.

If the gradient keeps pointing in the same direction:

    movement accumulates
         ↓
    faster optimization

If the gradient keeps changing direction:

    momentum helps reduce unnecessary oscillations.

## Normal Gradient Descent

Normal Gradient Descent updates the weight using:

    W_new = W_old - learning_rate × gradient

The current gradient completely determines the update.

For example:

    W = 5
    gradient = 2
    learning_rate = 0.1

    W_new = 5 - (0.1 × 2)

    W_new = 4.8

## Adding Momentum

Momentum introduces a velocity variable.

We calculate:

    velocity = momentum × previous_velocity
               + gradient

Then update the weight:

    W_new = W_old - learning_rate × velocity

The exact sign convention can vary between formulations,
but the important idea is:

    velocity remembers previous gradients.

Therefore the optimizer does not completely forget
what happened in previous steps.

## Numerical Example

Suppose:

    initial weight = 5
    learning rate = 0.1
    momentum = 0.9

Initial velocity:

    v = 0

Suppose the first gradient is:

    gradient = 2

Then:

    v = 0.9 × 0 + 2
    v = 2

Weight:

    W = 5 - 0.1 × 2
    W = 4.8

Now suppose the next gradient is:

    gradient = 2

Previous velocity:

    v = 2

New velocity:

    v = 0.9 × 2 + 2
    v = 3.8

Weight:

    W = 4.8 - 0.1 × 3.8
    W = 4.42

Notice something important:

The second update is larger because the optimizer
remembered the previous movement.

## Why Does Momentum Help?

Suppose gradients repeatedly point in the same direction:

    Gradient 1 → 2
    Gradient 2 → 2
    Gradient 3 → 2
    Gradient 4 → 2

Without momentum:

    update
    update
    update
    update

Each update depends only on the current gradient.

With momentum:

    previous movement
          +
    current gradient
          ↓
    larger velocity
          ↓
    faster movement

So momentum accelerates movement in directions
where the gradient consistently points.

In [1]:
import numpy as np

w = 5.0

learning_rate = 0.1
momentum = 0.9

velocity = 0.0

gradients = [2, 2, 2, 2, 2]

for step, gradient in enumerate(gradients, start=1):

    # Calculate velocity
    velocity = momentum * velocity + gradient

    # Update weight
    w = w - learning_rate * velocity

    print(
        f"Step {step}: "
        f"Gradient={gradient}, "
        f"Velocity={velocity:.4f}, "
        f"Weight={w:.4f}"
    )

Step 1: Gradient=2, Velocity=2.0000, Weight=4.8000
Step 2: Gradient=2, Velocity=3.8000, Weight=4.4200
Step 3: Gradient=2, Velocity=5.4200, Weight=3.8780
Step 4: Gradient=2, Velocity=6.8780, Weight=3.1902
Step 5: Gradient=2, Velocity=8.1902, Weight=2.3712


The velocity keeps increasing because the gradients
continue pointing in the same direction.

For example:

    Step 1
    velocity = 2

    Step 2
    velocity = 3.8

    Step 3
    velocity = 5.42

    Step 4
    velocity = 6.878

The optimizer is building up momentum.

This is why the weight can move faster than
ordinary Gradient Descent.

## Gradient Descent vs Momentum

### Normal Gradient Descent

    Current gradient
          ↓
    Weight update

It only uses the current gradient.

### Gradient Descent + Momentum

    Previous velocity
          +
    Current gradient
          ↓
    New velocity
          ↓
    Weight update

It remembers previous updates.